### PR TITLE
Change HTML attributes used for widgets to data- attributes

### PR DESCRIPTION
**Pull request by pichalite**

### CODE REVIEW

**Comment by pichalite** [comment]
Fixes #5435

**Comment by pitaj** [comment]
This should probably target develop, unless it isn't actually a breaking change.

**Comment by julianlam** [comment]
As widget areas are defined by themes, and themes will have to make this change in order for `.data-widget-area` to continue to match, this is a breaking change. Thanks @pitaj

**Comment by julianlam** [comment]
... and thanks @pichalite for the PR 😄

**Comment by RoiEXLab** [comment]
Thanks for the PR, I really appreciate it, that "my issue" is being adressed.
I have a small suggestion though...
Datasets are treated specially in javascript...
You would access datafields like this
```
<div id="id" data-very-important="secret"></div>
<script>
//Without jquery
console.log(document.getElementById("id").dataset.veryImportant);
//With jQuery 
console.log($("#id").data("veryImportant"));
</script>
```
TL;DR I'd recommend replacing `$this.attr('data-has-widget-class')` with `$this.data('hasWidgetClass')`.
The same applies of course for every other dataset

**Comment by julianlam** [comment]
To my knowledge, as there is no functional difference between using `.data('attr')` and `.attr('data-attr');`, I see no reason to change the code.

TIL, though.

**Comment by RoiEXLab** [comment]
No functional difference, correct.
It's shorter though... 
Without using jQuery it's a huge improvement though... The Attribute functions are just a pain to use...

**Comment by RoiEXLab** [comment]
Just mentioning: this PR isn't adressing 100% of my issue.
The component attribute still needs to be changed

**Comment by pichalite** [comment]
@RoiEXLab one step at a time.

Component attribute is not a simple thing to change. Requires lot of core changes and every single theme needs to be changed. Probably not going to happen anytime soon.

**Comment by RoiEXLab** [comment]
Of course!
You just said "Fixes #5435" and that's what made me sceptical

**Comment by pichalite** [comment]
You should probably move the component attribute change to a new issue.

**Comment by pitaj** [comment]
There is a difference between `.data('attr')` and `.attr('data-attr')`.

`.data` can't be used to set attribute values. It will set the internal jQuery data instead.
`.data` will fail if the attribute value is changed after the first call on that element.

Here's a JSFiddle demonstrating what I mean: https://jsfiddle.net/hdfumLw3/

**Comment by RoiEXLab** [comment]
I would call this a bug, and this is the first time I'm truely disappointed in jQuery... 
Worth to mention is also, that attr isn't affecting data either. Those 2 function are completely independent

**Comment by RoiEXLab** [comment]
This PR should probably added to the 1.5.0 milestone since it's a breaking change, just as a reminder...

**Comment by pitaj** [comment]
@RoiEXLab  We're not planning on introducing this change in 1.5.0; it'll probably wait until 2.0

**Comment by RoiEXLab** [comment]
In this case, all of #5435 should be adressed then...
Not necessarily in this PR...

**Comment by RoiEXLab** [comment]
@julianlam @barisusakli Seeing the recently removed deprecated routes:
Is there something that would speak against a 2-track approach, where using the current system would still work, but generate a warning, but the non deprecated way would require a data attribute?

**Comment by pitaj** [comment]
It's possible to do it with a deprecation, yes. Either way it'll be a lot of work. The main barrier is how many changes in core are necessary.

Also, it's not a pressing issue at it's not like browsers are going to be removing support any time soon.

**Comment by akhoury** [comment]
@julianlam during our conversation you mentioned that I should be making relevant PRs to Persona & Material themes, those are done, relevant PRs submitted, but is there any reason this won't affect Vanilla and Lavender? I clearly see the attrs in [there](https://github.com/NodeBB/nodebb-theme-lavender/search?q=widget-area&unscoped_q=widget-area) and [here](https://github.com/NodeBB/nodebb-theme-vanilla/search?q=widget-area&unscoped_q=widget-area)

it's good to note that `has-widget-class` and `has-widget-target ` didn't show up in any of the theme files I modified so far

**Comment by barisusakli** [comment]
>it's good to note that has-widget-class and has-widget-target didn't show up in any of the theme files I modified so far

Those two were probably removed when we refactored widgets to be rendered server side.

**Comment by julianlam** [comment]
In that case, you'll probably want to make the appropriate changes to vanilla and lavender as well.

Ideally we'd want to make this backwards compatible, but as the next version is breaking, I have no problems keeping this breaking.

Ain't PR review duty fun :smile:

**Comment by akhoury** [comment]
new pr backward compatible https://github.com/NodeBB/NodeBB/pull/7703
